File: sources/main/java/net/zxmushroom63/plugins/PluginLoader.js
window.pluginLoader = function pluginLoader(pluginsArr) {
  if (!window.eaglerPLoaderMainRun) {
    var searchParams = new URLSearchParams(location.search);
    searchParams.getAll("plugin").forEach((pluginToAdd) => {
      console.log(
        "EaglerPL: Adding plugin to loadlist from search params: " + pluginToAdd
      );
      pluginsArr.push(pluginToAdd);
    });
    if (
      !!eaglercraftXOpts &&
      !!eaglercraftXOpts.plugins &&
      Array.isArray(eaglercraftXOpts.plugins)
    ) {
      eaglercraftXOpts.plugins.forEach((pluginToAdd) => {
        console.log(
          "EaglerPL: Adding plugin to loadlist from eaglercraftXOpts: " +
            pluginToAdd
        );
        pluginsArr.push(pluginToAdd);
      });
    }
    window.eaglerPLoaderMainRun = true;
  }
  function checkPluginsLoaded(totalLoaded, identifier) {
    console.log(
      "EaglerPL: Checking if Plugins are finished :: " +
        totalLoaded +
        "/" +
        pluginsArr.length
    );
    if (totalLoaded >= pluginsArr.length) {
      clearInterval(identifier);
      window.pluginGracePeriod = false;
      console.log(
        "EaglerPL: Checking if Plugins are finished :: All plugins loaded! Grace period off."
      );
    }
  }
  function methodB(currentPlugin) {
    try {
      console.log("EaglerPL: Loading " + currentPlugin + " via method B.");
      var script = document.createElement("script");
      script.src = currentPlugin;
      script.setAttribute("data-plugin", currentPlugin);
      script.setAttribute("data-isplugin", true);
      script.onerror = () => {
        console.log(
          "EaglerPL: Failed to load " + currentPlugin + " via method B!"
        );
        script.remove();
        totalLoaded++;
      };
      script.onload = () => {
        console.log(
          "EaglerPL: Successfully loaded " + currentPlugin + " via method B."
        );
        totalLoaded++;
      };
      document.body.appendChild(script);
    } catch (error) {
      console.log(
        "EaglerPL: Oh no! The plugin " + currentPlugin + " failed to load!"
      );
      totalLoaded++;
    }
  }
  window.pluginGracePeriod = true;
  var totalLoaded = 0;
  var loaderCheckInterval = null;
  pluginsArr.forEach((c) => {
    let currentPlugin = c;
    console.log("EaglerPL: Starting " + currentPlugin);
    try {
      var req = new XMLHttpRequest();
      req.open("GET", currentPlugin);
      req.onload = function xhrLoadHandler() {
        console.log("EaglerPL: Loading " + currentPlugin + " via method A.");
        var script = document.createElement("script");
        try {
          script.src = "data:text/javascript;base64," + btoa(req.responseText);
        } catch (error) {
          methodB(currentPlugin);
          return;
        }
        script.setAttribute("data-plugin", currentPlugin);
        script.setAttribute("data-isplugin", true);
        script.onerror = () => {
          console.log(
            "EaglerPL: Failed to load " + currentPlugin + " via method A!"
          );
          script.remove();
          totalLoaded++;
        };
        script.onload = () => {
          console.log(
            "EaglerPL: Successfully loaded " + currentPlugin + " via method A."
          );
          totalLoaded++;
        };
        document.body.appendChild(script);
      };
      req.onerror = function xhrErrorHandler() {
        methodB(currentPlugin);
      };
      req.send();
    } catch (error) {
      methodB(currentPlugin);
    }
  });
  loaderCheckInterval = setInterval(() => {
    checkPluginsLoaded(totalLoaded, loaderCheckInterval);
  }, 500);
  console.log(
    "EaglerPL: Starting to load " + pluginsArr.length + " plugins..."
  );
};

File: sources/main/java/net/zxmushroom63/plugins/PluginLoader.java
package net.zxmushroom63.plugins;

import org.teavm.jso.JSBody;

public class PluginLoader {
    public static String[] plugins = {};

    @JSBody(params = { "plugins" }, script = "window.pluginLoader(plugins);")
    public static native void loadPlugins(String[] plugins);

    @JSBody(params = {}, script = "window.pluginLoader = function pluginLoader(pluginsArr) {\r\n" + //
            "  function checkPluginsLoaded(totalLoaded, identifier) {\r\n" + //
            "    console.log(\r\n" + //
            "      \"Checking if Plugins are finished :: \" +\r\n" + //
            "        totalLoaded +\r\n" + //
            "        \"/\" +\r\n" + //
            "        pluginsArr.length\r\n" + //
            "    );\r\n" + //
            "    if (totalLoaded >= pluginsArr.length) {\r\n" + //
            "      clearInterval(identifier);\r\n" + //
            "      window.pluginGracePeriod = false;\r\n" + //
            "      console.log(\r\n" + //
            "        \"Checking if Plugins are finished :: All plugins loaded! Grace period off.\"\r\n" + //
            "      );\r\n" + //
            "    }\r\n" + //
            "  }\r\n" + //
            "  function methodB(currentPlugin) {\r\n" + //
            "    try {\r\n" + //
            "      console.log(\"EaglerPL: Loading \" + currentPlugin + \" via method B.\");\r\n" + //
            "      var script = document.createElement(\"script\");\r\n" + //
            "      script.src = currentPlugin;\r\n" + //
            "      script.setAttribute(\"data-plugin\", currentPlugin);\r\n" + //
            "      script.setAttribute(\"data-isplugin\", true);\r\n" + //
            "      script.onerror = () => {\r\n" + //
            "        console.log(\r\n" + //
            "          \"EaglerPL: Failed to load \" + currentPlugin + \" via method B!\"\r\n" + //
            "        );\r\n" + //
            "        script.remove();\r\n" + //
            "        totalLoaded++;\r\n" + //
            "      };\r\n" + //
            "      script.onload = () => {\r\n" + //
            "        console.log(\r\n" + //
            "          \"EaglerPL: Successfully loaded \" + currentPlugin + \" via method B.\"\r\n" + //
            "        );\r\n" + //
            "        totalLoaded++;\r\n" + //
            "      };\r\n" + //
            "      document.body.appendChild(script);\r\n" + //
            "    } catch (error) {\r\n" + //
            "      console.log(\r\n" + //
            "        \"EaglerPL: Oh no! The plugin \" + currentPlugin + \" failed to load!\"\r\n" + //
            "      );\r\n" + //
            "      totalLoaded++;\r\n" + //
            "    }\r\n" + //
            "  }\r\n" + //
            "  window.pluginGracePeriod = true;\r\n" + //
            "  var totalLoaded = 0;\r\n" + //
            "  var loaderCheckInterval = null;\r\n" + //
            "  pluginsArr.forEach((c) => {\r\n" + //
            "    let currentPlugin = c;\r\n" + //
            "    console.log(\"EaglerPL: Starting \" + currentPlugin);\r\n" + //
            "    try {\r\n" + //
            "      var req = new XMLHttpRequest();\r\n" + //
            "      req.open(\"GET\", currentPlugin);\r\n" + //
            "      req.onload = function xhrLoadHandler() {\r\n" + //
            "        console.log(\"EaglerPL: Loading \" + currentPlugin + \" via method A.\");\r\n" + //
            "        var script = document.createElement(\"script\");\r\n" + //
            "        script.src = \"data:text/javascript;base64,\" + btoa(req.responseText);\r\n" + //
            "        script.setAttribute(\"data-plugin\", currentPlugin);\r\n" + //
            "        script.setAttribute(\"data-isplugin\", true);\r\n" + //
            "        script.onerror = () => {\r\n" + //
            "          console.log(\r\n" + //
            "            \"EaglerPL: Failed to load \" + currentPlugin + \" via method A!\"\r\n" + //
            "          );\r\n" + //
            "          script.remove();\r\n" + //
            "          totalLoaded++;\r\n" + //
            "        };\r\n" + //
            "        script.onload = () => {\r\n" + //
            "          console.log(\r\n" + //
            "            \"EaglerPL: Successfully loaded \" + currentPlugin + \" via method A.\"\r\n" + //
            "          );\r\n" + //
            "          totalLoaded++;\r\n" + //
            "        };\r\n" + //
            "        document.body.appendChild(script);\r\n" + //
            "      };\r\n" + //
            "      req.onerror = function xhrErrorHandler() {\r\n" + //
            "        methodB(currentPlugin);\r\n" + //
            "      };\r\n" + //
            "      req.send();\r\n" + //
            "    } catch (error) {\r\n" + //
            "      methodB(currentPlugin);\r\n" + //
            "    }\r\n" + //
            "  });\r\n" + //
            "  loaderCheckInterval = setInterval(() => {\r\n" + //
            "    checkPluginsLoaded(totalLoaded, loaderCheckInterval);\r\n" + //
            "  }, 500);\r\n" + //
            "  console.log(\r\n" + //
            "    \"EaglerPL: Starting to load \" + pluginsArr.length + \" plugins...\"\r\n" + //
            "  );\r\n" + //
            "};\r\n" + //
            "")
    public static native void loadLoader();

    @JSBody(params = { "plugins" }, script = "localStorage.setItem('pl::plugins', JSON.stringify(plugins))")
    private static native void savePlugins(String[] plugins);

    @JSBody(params = {}, script = "try { return JSON.parse(localStorage.getItem('pl::plugins')||'[]') } catch(err) {return []}")
    private static native String[] retrievePlugins();

    public static void savePluginsToLocalStorage() {
        savePlugins(plugins);
    };

    public static void loadPluginsFromLocalStorage() {
        plugins = retrievePlugins();
    };
}